internal/cache/live_query.go

```diff
@@ -0,0 +1,319 @@
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	"github.com/upbound/xgql/internal/clients"
+	"github.com/upbound/xgql/internal/live_query"
+)
+
+// WithLiveQueries wraps NewCacheFn with a cache.Cache that tracks objects
+// and object lists and notifies the live query in request context of changes.
+func WithLiveQueries(fn clients.NewCacheFn) clients.NewCacheFn {
+	return func(cfg *rest.Config, o cache.Options) (cache.Cache, error) {
+		c, err := fn(cfg, o)
+		if err != nil {
+			return nil, err
+		}
+		return &liveQueryCache{
+			Cache:   c,
+			scheme:  o.Scheme,
+			queries: make(map[uint64]*liveQueryTracker),
+			handles: make(set[schema.GroupVersionKind]),
+		}, nil
+	}
+}
+
+var _ toolscache.ResourceEventHandler = (*liveQueryCache)(nil)
+
+// liveQueryCache is a cache.Cache that registers cache.Informer listeners for any
+// retrieved object if executed in the context of a live query. When liveQueryCache
+// is notified of events, it will trigger any active live queries.
+type liveQueryCache struct {
+	cache.Cache
+	scheme *runtime.Scheme
+
+	lock    sync.Mutex
+	queries map[uint64]*liveQueryTracker
+	handles set[schema.GroupVersionKind]
+}
+
+// Get implements cache.Cache. It wraps an underlying cache.Cache and sets up an Informer
+// event handler that marks current live query as dirty if the current context has a live query.
+func (c *liveQueryCache) Get(ctx context.Context, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error {
+	if err := c.Cache.Get(ctx, key, obj, opts...); err != nil {
+		return err
+	}
+	return c.trackObject(ctx, obj)
+}
+
+// List implements cache.Cache. It wraps an underlying cache.Cache and sets up an Informer
+// event handler that marks current live query as dirty if the current context has a live query.
+func (c *liveQueryCache) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	if err := c.Cache.List(ctx, list, opts...); err != nil {
+		return err
+	}
+	return c.trackObject(ctx, list)
+}
+
+// trackObject registers object or object list with a tracker for the live query.
+// any updated from cache.Informer is broadcast to all live query trackers, if the
+// changed object is tracked by a given liveQueryTracker, the live query associated
+// with the tracker is Trigger()'d.
+func (c *liveQueryCache) trackObject(ctx context.Context, object runtime.Object) error {
+	qid, ok := live_query.IsLive(ctx)
+	// if this isn't a live query context, skip.
+	if !ok {
+		return nil
+	}
+	gvk, err := apiutil.GVKForObject(object, c.scheme)
+	if err != nil {
+		return err
+	}
+	if _, ok := object.(client.ObjectList); ok {
+		// We need the non-list GVK, so chop off the "List" from the end of the kind.
+		gvk.Kind = strings.TrimSuffix(gvk.Kind, "List")
+	}
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	// register event handler for the GVK that if we aren't watching it already.
+	if !c.handles.Contains(gvk) {
+		i, err := c.getInformer(ctx, object, gvk)
```

Comment:
this waits for synced, right? I.e. blocking on this is not super helpful. It can take seconds.

Reply:
You're right. This call happens after a get or list of the same gvk from the wrapped cache, so it has synced by that point.

Reply:
was wondering which impact holding the `c.lock` has for e.g. 10s, from the user experience point of view.

Reply:
Especially if there multiple resources being watched, and this essentially linearizes everything.

Reply:
The performance is comparable to the original implementation. The informer sync has already happened by this point since we're tracking after list or get from cache succeeded.

Reply:
I can get the informer outside without holding the lock.

Reply:
done